bin/main.rb

```diff
@@ -1,58 +1,57 @@
+require_relative '../lib/board.rb'
+require_relative '../lib/player.rb'
 
-def display_grid(board)
-  board.each do |n|
-    n.each do |j|
-      print "#{j} "
-    end
-    print "\n"
-  end
+player1 = Player.new
+player2 = Player.new
+puts "Welcome to tic-tac-toe game!\n\n"
+board = Board.new
+puts 'Enter Player-1 name:'
+name1 = gets.chomp
+until player1.valid_name?(name1)
+  puts 'Enter a valid name'
+  name1 = gets.chomp
 end
-
-game_on = true
-colors = %w[X O]
-puts 'Welcome to tic-tac-toe game'
-display_grid(board)
-puts "\nEnter Player-1 name:"
-player1 = gets.chomp
 puts "\nPlease select X or O"
 color1 = gets.chomp
-puts "\n#{player1}, your color is #{color1}"
-colors.delete(color1)
+until %w[O X].include?(color1)
+  puts "\nEnter a valid symbol"
+  color1 = gets.chomp
+end
+color2 = board.get_color(color1)
 puts "\nEnter Player-2 name:"
-player2 = gets.chomp
-color2 = colors[0]
-puts "\n#{player2}, your color is #{color2}"
-
-winner1 = false
-winner2 = false
-positions = { '1' => [0, 0], '2' => [1, 0], '3' => [2, 0],
-              '4' => [0, 1], '5' => [1, 1], '6' => [2, 1],
-              '7' => [0, 2], '8' => [1, 2], '9' => [2, 2] }
-
-while game_on
-  9.times do
-    puts "\n#{player1}, please input your move (1-9):"
-    move = gets.chomp
-    if /[0-9]/.match(move).nil?
-      puts "\nThis is not a valid input, try again"
-    elsif board[positions[move][0]][positions[move][1]] == 'X' or board[positions[move][0]][positions[move][1]] == 'O'
-      puts "\nThis position is already taken, try again"
-    else
-      board[positions[move][0]][positions[move][1]] = color1
-    end
-    display_grid(board)
-  end
-  game_on = false
-  winner1 = true
+name2 = gets.chomp
+until player2.valid_name?(name2)
+  puts "\nEnter a valid name"
+  name2 = gets.chomp
 end
+puts "\nHello #{name1}, your symbol is #{color1}!"
+puts "\nHello #{name2}, your symbol is #{color2}!\n\n"
+count = 0
+print board.display_grid + "\n"
+loop do
+  puts "#{name1} select a Position (#{color1})"
+  player1_pos = gets.chomp.to_i
+  until board.valid_move?(player1_pos)
+    puts "\nEnter a valid move (#{color1})\n\n"
+    print board.display_grid + "\n"
+    player1_pos = gets.chomp.to_i
+  end
+  count += 1
+  puts board.display_grid(player1_pos, color1)
+  return puts "\n#{name1} won!" if player1.wins?(player1_pos)
+  return puts "\nMatch Draw!!" if count == 9
 
-if winner1
-  puts "Game Over, #{player1} won, congratulations!"
-elsif winner2
-  puts "Game Over, #{player2} won, congratulations!"
-else
-  puts "Game Over, it's a tie, too bad!"
+  puts "#{name2} select the position (#{color2})"
+  player2_pos = gets.chomp.to_i
+  until board.valid_move?(player2_pos)
+    puts "\nEnter a valid move (#{color2})\n\n"
+    print board.display_grid + "\n"
+    player2_pos = gets.chomp.to_i
+  end
+  count += 1
+  puts board.display_grid(player2_pos, color2)
+  return puts "\n#{name2} won!" if player2.wins?(player2_pos)
+  return puts "\nMatch Draw!!" if count == 9
 end
```

Comment:
* [ ] A loop is needed in this game (recommended is `while` but anything else will do) ([OdinProject hint](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone2/loop.png))
* [ ] Inside loop, all actions should be repeated for each player separately
* [ ] A loop needs to be terminated at some point (after a winning move or draw to be more specific). It can be done with variable:

- **[SUGGESTION]** Since you changed the code related to the game loop, I suggest that you implement you game loop as below:
```ruby
game_on = true
current_player = playerOne
# initial instructions for player

while game_on
  # loop for each move
   # display grid and UI messages for current player
   # get input from current player
   # check if it's a valid input
   # if valid_move? check for win/draw/keep playing status
  if winner # and/or draw (the exact condition in this milestone is not important)
    game_on = false
  end
  # display messages if the game is finished
 # switch current_player to the other player if the game is not finished
end
```

Reply:
Good morning, Ahmed!
So, we didn't understand this review regarding the loop part, because we are using a loop (not a while loop, but it is a loop) that ends when a winner or a draw is returned inside the conditionals. We can implement it with a while loop as well, but our code is working fine with that loop. We are changing the observation about attributes. Thank you!